Expand documentation of Set

jakobnissen opened the pull request:
Hash sets have properties that can be counter-intuitive to people not familiar
with computer science: They provide constant-time membership testing, are unordered,
and deduplicate their elements as determined by `isequal` and `hash`.

This commit expands the documentation of `Set` to mention the above properties,
and also that it is a subtype of `AbstractSet`.

jakobnissen committed:
Expand documentation of Set

Hash sets have properties that can be counter-intuitive to people not familiar
with computer science: They provide constant-time membership testing, are unordered,
and deduplicate their elements as determined by `isequal` and `hash`.

This commit expands the documentation of `Set` to mention the above properties,
and also that it is a subtype of `AbstractSet`.

petvana commented on base/set.jl:
```diff
@@ -1,5 +1,33 @@
+
+`Set`s are containers that provide fast lookup and membership testing.
+Elements in a `Set` are unique, as determined by the element's implementation of
+`isequal` and `hash`. Adding an element that already is in a `Set` does nothing.
```
This comment may be misleading. How to define an element that is already in a `Set`? By `isequal` or `in`, respectively? Then you can replace `1` by `1.0`, so the adding (pushing into) does something, but the number of elements remains the same.

``` julia
julia> s = Set()
Set{Any}()

julia> push!(s, 1)
Set{Any} with 1 element:
  1

julia> push!(s, 1.0)
Set{Any} with 1 element:
  1.0

```

jakobnissen replied:
You are right, this was wrong. I've removed the statement, since the actual behaviour of adding two objects with the same hash is presumably an implementation detail.

petvana replied:
I can't find this corner case to be documented. So what about this?
"Adding an element that already is `in` the `Set` replaces the old element."

mcabbott replied:
An example showing that `push!` is how you add elements might also be good. So far the fact that the Set is mutable isn't described or shown.

jakobnissen committed:
Amend docs for Set element uniqueness

mcabbott commented on base/set.jl:
```diff
@@ -1,5 +1,33 @@
+
+See also: [`AbstractSet`](@ref), [`BitSet`](@ref), [`Dict`](@ref)
+
+# Examples
```
Re examples, it would be nice to show `in`, and to show some example that would fail if the check were `==` or `===` not `isequal`. Maybe:
```
julia> 0.0 in Set([-0.0, NaN, Inf])  # 0.0 == -0.0 is true, but Set uses isequal
false

julia> [2] in Set(([1], [2], [3]))  # [2] === [2] is false, but Set uses isequal
true
```

jakobnissen replied:
In examples, added `0` vs `0.0` vs `-0.0`.

mcabbott commented on base/set.jl:
```diff
@@ -1,5 +1,33 @@
+"""
+	Set{T} <: AbstractSet{T}
+
+The default implementation of hash sets.
```
"hash set" seems a little technical for the first line. Can it just explain somehow, that membership is the only concept? Explain `isequal` and `hash` later.

KristofferC replied:
To me, this sentence doesn't really add anything that the text below already explains.

KristofferC commented on base/set.jl:
```diff
@@ -1,5 +1,33 @@
+Set{Int64} with 1 element:
+  1
+
+julia> Set(['a', 'b', 'c'])
```
```suggestion
julia> Set(['a', 'b', 'c', 'a'])
```

to show the uniqueness?

KristofferC commented on base/set.jl:
```diff
@@ -19,17 +47,7 @@ function Set{T}(s::KeySet{T, <:Dict{T}}) where {T}
-empty set. Should be used instead of [`BitSet`](@ref) for sparse integer sets, or
-for sets of arbitrary objects.
-
-See also: [`push!`](@ref), [`empty!`](@ref), [`union!`](@ref), [`in`](@ref).
```
These references were pretty good. Why remove?

KristofferC commented on base/set.jl:
````diff
@@ -1,5 +1,33 @@
+
+# Examples
+```jldoctest filter = r"^Set{[A-Za-z0-9]+} with [0-9]+ element(s)?:"
+julia> Set()
````
This and the example below feel kind of pointless in that they don't really show any interesting property of what it is to be a `Set`.

jakobnissen committed:
Address reviewer comments:

* Remove redundant first line
* Re-add removed "See also" references
* Removed redundant empty Set instantiation example
* Have example show mutability
* Have example show deduplication
* Have example show behaviour under `isequal`

jakobnissen commented:
I've addressed your comments with the following changes
* Remove redundant first line
* Re-add removed "See also" references
* Removed redundant empty Set instantiation example
* Have example show mutability
* Have example show deduplication
* Have example show behaviour under `isequal`

petvana commented on base/set.jl:
```diff
@@ -1,5 +1,35 @@
+  0.0
+  -0.0
+
+julia> push!(s, 'a'); push!(s, 0.0)
```
What about pushing `0` to better highlight the mutability?
```suggestion
julia> push!(s, 'a'); push!(s, 0)
```
... and than `0.0 in s`

jakobnissen replied:
The issue with that is that I think the current behaviour where `0` overwrites `0.0` is an implementation detail (i.e. I think keeping `0.0` would be documented behaviour, too), so it's better to not have it in the doc example.

jakobnissen committed:
Fix doctest mistake

Moelf commented:
can we finally close #23269 while at it?

jakobnissen committed:
Update Set example

jakobnissen commented on base/set.jl:
```diff
@@ -14,14 +14,14 @@ See also: [`AbstractSet`](@ref), [`BitSet`](@ref), [`Dict`](@ref),
 julia> s = Set(Any[0, -0.0, 0.0])
 Set{Any} with 2 elements:
   0.0
   -0.0
```
See the updated example. The `in` examples show `==` vs `===` vs `isequal`, but the first example is simpler.

mcabbott replied:
Ok. But are we constrained to one example? I suggested a tuple as one way to indicate that it need not always be a vector, or any particular type.

mcabbott replied:
I guess my vote is that 2 examples could be enough. One for the most basic "unique items, not in order, mutable" idea. The other showing perhaps surprising Float64 behaviour, to point out that it's not `==`. 

The mixed Int+Float64 behaviour seems extra weird to me, and not so useful, I'd vote to leave that out. 

Extended help does seem a bit like overkill to me. Once you have one pointer that it's about `isequal`, then you can read up on that if you care. But trying to repeat all the details here (and there are more, esp. about `missing`) seems too much; rather say all of that clearly in one place.

jakobnissen committed:
Add extended help

jakobnissen commented:
I added more examples by recommendation - if this docstring is too long let me know.

jakobnissen committed:
Fix whitespace

jakobnissen committed:
Reduce Set docstring

petvana commented on base/set.jl:
````diff
@@ -1,5 +1,42 @@
+
+# Examples
+```jldoctest filter = r"^\\S.+"
+julia> s = Set("aaBca")
````
```suggestion
julia> s = Set("aaBca"); push!(s, 'b')
```
If you are worried it to be too long, I guess you can connect these together without losing any vital information.

KristofferC commented on base/set.jl:
```diff
@@ -1,5 +1,42 @@
+
+`Set`s have efficient implementations of set operations such as `in`, `union` and `intersect`.
+Elements in a `Set` are unique, as determined by the elements' definitions of
+`isequal` and `hash`.
```
I don't really get the addition of `hash` here. You could have a set with a bunch of elements that all hash to `1`, it would just be inefficient.

jakobnissen replied:
The idea is to prevent a user overloading `isequal` without overloading `hash` (e,g, having a bunch of not-equal elements that hash to `1` is an implementation error). This is not legal according to `isequal` docs though, so it might be redundant to also state it here. I'll remove it.

petvana replied:
Here, the problem is kind of the opposite, i.e., having multiple `isequal` elements with different hashes. So, omitting `hash` may easily lead to hard-to-debug bugs, like
``` julia
julia> import Base: isequal, hash

julia> struct X; x; end

julia> isequal(X(0.0), X(0))
false

julia> isequal(a::X, b::X) = a.x == b.x;

julia> isequal(X(0.0), X(0))
true

julia> s = Set([X(0.0), X(0)]) # Incorrect
Set{X} with 2 elements:
  X(0)
  X(0.0)

julia> hash.([X(0.0), X(0)])
2-element Vector{UInt64}:
 0x29f6c96c4b9e8bb8
 0x0792f7046dd5cf45

julia> hash(x::X) = hash(x.x);

julia> s = Set([X(0.0), X(0)]) # Correct
Set{X} with 1 element:
  X(0)
```
Btw, `hash` is usually mention for other languages too. Just to name few: [C++](https://en.cppreference.com/w/cpp/container/unordered_set), [Rust](https://doc.rust-lang.org/std/collections/hash_set/struct.HashSet.html), [Java](https://docs.oracle.com/javase/8/docs/api/java/util/Set.html) ... unless the set is BST-based, of course.

petvana replied:
`isequal` is actually called to check `hash` collisions for `Dict` (which implements `Set` internally) in
https://github.com/JuliaLang/julia/blob/0062c2653dcddcbf9adb80d23b0ff5eef72f7802/base/dict.jl#L273

mcabbott replied:
Ah ok. So it does depend on both. 

Maybe the description should first say what the contract is (unique according to `isequal`, just like `unique`) and later say that the fast implementation needs `hash`? Uniqueness in the 1st paragraph, what's efficient in the second?

jakobnissen replied:
Is that really necessary for the docstring, though? To me, mentioning `hash` seem like an optimization based on the implementation detail of Set, and that is usually outside the scope of a docstring.

petvana replied:
IMHO, yes! These two functions are entangled tightly. So, mentioning only one seems dangerous to me because it motivates the user to re-define `isequal` only, and it may lead to an incorrect/undefined behavior.

jakobnissen committed:
Remove mention of element hashing

jakobnissen commented:
OK so it's clear to me that we don't all exactly agree on what should and shouldn't go in the docstring. But let's not let the perfect be the enemy of the good and let's make sure this PR doesn't stall out and get nowhere due to bikeshedding.

aviatesk reviewed (approved):
LGTM. Any objections to this? Otherwise I'd like to merge this tmrw.
